bip-unknown-var-budget-script.mediawiki

```diff
@@ -0,0 +1,722 @@
+
+Negative numbers are not natively supported in v2 Tapscript.  Arbitrary precision makes them difficult to manipulate and negative values are not used meaningfully in bitcoin transactions.
+
+===Enabled Opcodes===
+
+Fifteen opcodes which were removed in v0.3.1 are re-enabled in v2 Tapscript.
+
+If there are less than the required number of stack elements, these opcodes fail.  Equivalently, a requirement to pop off the stack which cannot be satisfied causes the opcode to fail.
+
+Splice opcodes:
+
+{|
+! Opcode
+! Value
+! Required Stack Elements
+! Definition
+|-
+|OP_CAT
+|126
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# Append B to A.
+# Push A onto the stack.
+|-
+|OP_SUBSTR
+|127
+|3
+|
+# Pop LEN off the stack.
+# Pop BEGIN off the stack.
+# Pop A off the stack.
+# Remove BEGIN bytes from the front of A (all bytes if BEGIN greater than length of A).
+# If length(A) is greater than value(LEN), truncate A to length value(LEN).
+# Push A onto the stack.
+|-
+|OP_LEFT
+|128
+|2
+|
+# Pop OFFSET off the stack.
+# Pop A off the stack.
+# If length(A) is greater than value(OFFSET), truncate A to length value(OFFSET).
+# Push A onto the stack.
+|-
+|OP_RIGHT
+|129
+|2
+|
+# Pop OFFSET off the stack.
+# Pop A off the stack.
+# Remove value(OFFSET) bytes from the front of A, or all bytes if value(OFFSET) is greater than the length(A).
+# Push A onto the stack.
+|}
+
+Bit operation opcodes:
+
+{|
+! Opcode
+! Value
+! Required Stack Elements
+! Definition
+|-
+|OP_INVERT
+|131
+|1
+|
+# Pop A off the stack.
+# For each byte in A, replace it with that byte bitwise XOR 0xFF (i.e. invert the bits)
+# Push A onto the stack.
+|-
+|OP_AND
+|132
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# If B is longer than A, swap B and A.
+# For each byte in A (the longer operand): bitwise AND it with the equivalent byte in B (or 0 if past end of B)
+# Push A onto the stack.
+|-
+|OP_OR
+|133
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# If B is longer than A, swap B and A.
+# For each byte in B (the shorter operand): bitwise OR it with the equivalent byte in A.
+# Push A onto the stack.
+|-
+|OP_XOR
+|134
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# If B is longer than A, swap B and A.
+# For each byte in B (the shorter operand): exclusive OR it with the equivalent byte in A.
+# Push A onto the stack.
+|}
+
+Bitshift opcodes:
+
+Note that these are raw bitshifts, unlike the sign-preserving arithmetic shifts in Bitcoin v0.3.0, and as such they also do not truncated trailing zeroes from results: they are renamed OP_UPSHIFT (nee OP_LSHIFT) and OP_DOWNSHIFT (nee OP_RSHIFT).
+
+Note also that OP_UPSHIFT can produce huge results, and so must be checked for limits prior to evaluation.  It is also carefully defined to avoid reallocating twice (reallocating to prepend bytes, then again to append a single byte) which has the practical advantage of being about to share the same downward bitshift routine as OP_DOWNSHIFT.
```

Comment:
> which has the practical advantage of being about to share the same downward bitshift routine as OP_DOWNSHIFT.

it seems like there's a typo in here messing with the grammar of the sentence but I can't put my finger on it

Reply:
!

....being *able* to share...